### PR TITLE
Unused struct fields

### LINKED ISSUE #533: Missing location info in the hir of struct field
Missing location info in the hir of struct field

### PR DESCRIPTION
**Pull request by thomasyonug**
1. make struct field carry the location info. Fixed #533 
2. detect unused struct field.

### CODE REVIEW

**Comment by thomasyonug** [comment]
This pr can be merged after #532

**Comment by thomasyonug** [comment]
The behavior of ci is weird, i pass all the test in my local env.

**Comment by philberty** [comment]
I'm getting:

```
                === rust Summary ===

# of expected passes            2901
# of unexpected failures        63
# of expected failures          15
```

I think a good place to start would be to look at the warnings:

```
g++  -fno-PIE -c  -DIN_GCC_FRONTEND -g -O0 -DIN_GCC     -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variad
ic-macros -Wno-overlength-strings -fno-common  -DHAVE_CONFIG_H -I. -Irust -I../../gccrs/gcc -I../../gccrs/gcc/rust -I../../gccrs/gcc/../include -I../../gccrs/gcc/../libcpp/include -I../../gccrs/gcc/../libcody  -I../../gccrs/gcc/../libdecnumber -I.
./../gccrs/gcc/../libdecnumber/bid -I../libdecnumber -I../../gccrs/gcc/../libbacktrace   -o rust/rust-hir-full-test.o -MT rust/rust-hir-full-test.o -MMD -MP -MF rust/.deps/rust-hir-full-test.TPo -std=c++11 -Wno-unused-parameter -Werror=overloaded-
virtual -I ../../gccrs/gcc/rust -I ../../gccrs/gcc/rust/lex -I ../../gccrs/gcc/rust/parse -I ../../gccrs/gcc/rust/ast -I ../../gccrs/gcc/rust/analysis -I ../../gccrs/gcc/rust/backend -I ../../gccrs/gcc/rust/expand -I ../../gccrs/gcc/rust/hir/tree 
-I ../../gccrs/gcc/rust/hir -I ../../gccrs/gcc/rust/resolve -I ../../gccrs/gcc/rust/util -I ../../gccrs/gcc/rust/typecheck -I ../../gccrs/gcc/rust/lint ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc                                            
In file included from ../../gccrs/gcc/rust/ast/rust-ast-full.h:24,                                                                                                                                                                                     
                 from ../../gccrs/gcc/rust/hir/tree/rust-hir-full-test.cc:19:                                                                                                                                                                          
../../gccrs/gcc/rust/ast/rust-item.h: In constructor ‘Rust::AST::StructField::StructField(Rust::Identifier, std::unique_ptr<Rust::AST::Type>, Rust::AST::Visibility, Location, std::vector<Rust::AST::Attribute>)’:                                    
../../gccrs/gcc/rust/ast/rust-item.h:1846:10: warning: ‘Rust::AST::StructField::node_id’ will be initialized after [-Wreorder]
 1846 |   NodeId node_id;                                                                                                                                                                                                                              
      |          ^~~~~~~                                                                                                                                                                                                                               
../../gccrs/gcc/rust/ast/rust-item.h:1844:12: warning:   ‘Location Rust::AST::StructField::locus’ [-Wreorder]                                                                                                                                          
 1844 |   Location locus;                                                                                                                                                                                                                              
      |            ^~~~~                                                                                                                                                                                                                               
../../gccrs/gcc/rust/ast/rust-item.h:1855:3: warning:   when initialized here [-Wreorder]                                                                                                                                                              
 1855 |   StructField (Identifier field_name, std::unique_ptr<Type> field_type,                                                                                                                                                                        
      |   ^~~~~~~~~~~                                                                                                                                                                                                                                  
../../gccrs/gcc/rust/ast/rust-item.h: In copy constructor ‘Rust::AST::StructField::StructField(const Rust::AST::StructField&)’:                                                                                                                        
../../gccrs/gcc/rust/ast/rust-item.h:1846:10: warning: ‘Rust::AST::StructField::node_id’ will be initialized after [-Wreorder]                                                                                                                         
 1846 |   NodeId node_id;                                                                                                                                                                                                                              
      |          ^~~~~~~                                                                                                                                                                                                                               
../../gccrs/gcc/rust/ast/rust-item.h:1844:12: warning:   ‘Location Rust::AST::StructField::locus’ [-Wreorder]                                                                                                                                          
 1844 |   Location locus;                                                                                                                                                                                                                              
      |            ^~~~~                                                                                                                                                                                                                               
../../gccrs/gcc/rust/ast/rust-item.h:1864:3: warning:   when initialized here [-Wreorder]                                                                                                                                                              
 1864 |   StructField (StructField const &other)                                       
```

**Comment by philberty** [comment]
In fact it might be a good idea to split the commit which adds the location info to the struct field to be a separate  PR to we can isolate which commit breaks the tests.

**Comment by thomasyonug** [comment]
> In fact it might be a good idea to split the commit which adds the location info to the struct field to be a separate PR to we can isolate which commit breaks the tests.

Good idea!

**Comment by thomasyonug** [comment]
> The behavior of ci is weird, i pass all the test in my local env. Let me have a look into it

The root cause is I forgot to write a return stmt in an iterate callback. It's a really really place point to start with warnings.
Thanks @philberty .

File: gcc/rust/lint/rust-lint-scan-deadcode.h [hunk 1]
```diff
@@ -46,15 +46,34 @@ class ScanDeadcode : public MarkLiveBase
   void visit (HIR::Function &function) override
   {
     HirId hirId = function.get_mappings ().get_hirid ();
-    warning (hirId, function.get_locus (), "function",
-	     function.get_function_name (), "used");
+    if (should_warn (hirId))
+      {
+	warning (function.get_locus (), "function",
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-scan-deadcode.h:51]: I don't personally like this wrapper around the warning. I would prefer each of the warnings were done by calling rust_warning_at directly.

File: gcc/rust/lint/rust-lint-scan-deadcode.h [hunk 2]
```diff
@@ -46,15 +46,34 @@ class ScanDeadcode : public MarkLiveBase
...
   }
 
   void visit (HIR::StructStruct &stct) override
   {
     HirId hirId = stct.get_mappings ().get_hirid ();
-    warning (hirId, stct.get_locus (), "struct", stct.get_identifier (),
-	     "constructed");
+    if (should_warn (hirId))
+      {
+	warning (stct.get_locus (), "struct", stct.get_identifier (),
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-scan-deadcode.h:61]: same here

File: gcc/rust/lint/rust-lint-scan-deadcode.h [hunk 3]
```diff
@@ -46,15 +46,34 @@ class ScanDeadcode : public MarkLiveBase
...
+		 "constructed");
+      }
+    else
+      {
+	// only warn the unused fields when in unwarned struct.
+	stct.iterate ([&] (HIR::StructField &field) -> bool {
+	  HirId field_hir_id = field.get_mappings ().get_hirid ();
+	  if (should_warn (field_hir_id))
+	    {
+	      warning (field.get_locus (), "field", field.get_field_name (),
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-scan-deadcode.h:71]: and here

File: gcc/rust/lint/rust-lint-scan-deadcode.h [hunk 4]
```diff
@@ -64,15 +83,16 @@ class ScanDeadcode : public MarkLiveBase
   ScanDeadcode (std::set<HirId> &live_symbols)
     : live_symbols (live_symbols), resolver (Resolver::Resolver::get ()){};
 
-  void warning (HirId hirId, Location loc, const std::string &span,
-		const std::string &name, const std::string &participle) const
+  void warning (Location loc, const std::string &span, const std::string &name,
```
Comment (philberty) [inline gcc/rust/lint/rust-lint-scan-deadcode.h:86]: Lets get rid of this function and just call rust_warning_at directly.

**Review by philberty — CHANGES_REQUESTED** [review]
I think the rest of this looks ok apart from how you emit the warning

File: gcc/rust/lint/rust-lint-scan-deadcode.h (hunk 1 above)
Comment (thomasyonug) [reply gcc/rust/lint/rust-lint-scan-deadcode.h:51]: Ok, gotcha

**Review by philberty — APPROVED** [review]

**Pull request merged**